src/backend/distributed/connection/placement_connection.c

```diff
@@ -229,272 +222,373 @@ GetPlacementConnection(uint32 flags, ShardPlacement *placement, const char *user
 	}
 
-	return returnConnectionReference->connection;
+	return StartPlacementListConnection(flags, list_make1(placementAccess), userName);
 }
 
 
 /*
- * CheckExistingPlacementConnections check whether any of the existing
- * connections is usable. If so, return it, otherwise return NULL.
+ * GetPlacementListConnection establishes a connection for a set of placement
+ * accesses.
  *
- * A connection is usable if it is not in use, the user matches, DDL/DML usage
- * match and cached connection are allowed.  If no existing connection
- * matches, but a new connection would conflict (e.g. a connection already
- * exists but isn't usable, and the desired connection needs to execute
- * DML/DML) an error is thrown.
+ * See StartPlacementListConnection for details.
+ */
+MultiConnection *
+GetPlacementListConnection(uint32 flags, List *placementAccessList, const char *userName)
+{
+	MultiConnection *connection = StartPlacementListConnection(flags, placementAccessList,
+															   userName);
+
+	FinishConnectionEstablishment(connection);
+	return connection;
+}
+
+
+/*
+ * StartPlacementListConnection returns a connection to a remote node suitable for
+ * a placement accesses (SELECT, DML, DDL) or throws an error if no suitable
+ * connection can be established if would cause a self-deadlock or consistency
+ * violation.
  */
-static ConnectionReference *
-CheckExistingPlacementConnections(uint32 flags,
-								  ConnectionPlacementHashEntry *placementEntry,
-								  const char *userName)
+MultiConnection *
+StartPlacementListConnection(uint32 flags, List *placementAccessList,
+							 const char *userName)
 {
-	dlist_iter it;
+	char *freeUserName = NULL;
+	bool foundModifyingConnection = false;
+	ListCell *placementAccessCell = NULL;
+	List *placementEntryList = NIL;
+	ListCell *placementEntryCell = NULL;
+	MultiConnection *connection = NULL;
+	ShardPlacement *lastPlacement = NULL;
+
+	if (userName == NULL)
+	{
+		userName = freeUserName = CurrentUserName();
+	}
 
 	/*
-	 * If there's a connection that has executed DML/DDL, always return it if
-	 * possible. That's because we only can execute DML/DDL over that
-	 * connection.  Checking this first also allows us to detect conflicts due
-	 * to opening a second connection that wants to execute DML/DDL.
+	 * Go through all placement accesses to find a suitable connection.
+	 *
+	 * If none of the placements have been accessed in this transaction, connection
+	 * remains NULL.
+	 *
+	 * If one or more of the placements have been modified in this transaction, then
+	 * use the connection that performed the write. If placements have been written
+	 * over multiple connections or the connection is not available, error out.
+	 *
+	 * If placements have only been read in this transaction, then use the last
+	 * suitable connection found for a placement in the placementAccessList.
 	 */
-	if (placementEntry->modifyingConnection)
+	foreach(placementAccessCell, placementAccessList)
```

Comment:
Not sure that's a good idea. There could well be multiple connections in a single statement outside of a transaction block (e.g. in a UDF call). Fortunately, iterating through a list of size 1 is probably not a performance bottleneck. :)